### PR TITLE
Fix node allocatable resource validation

### PR DESCRIPTION
**Pull request by NickrenREN**
GetNodeAllocatableReservation gets all the reserved resource value
Allocatable resource = capacity - reservation

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by NickrenREN** [comment]
ping @jingxu97  @vishh

**Comment by jingxu97** [comment]
@NickrenREN Thank you for finding and fixing the bug! But a problem I see is getNodeAllocatableAbsolute sets the value to 0 if it is smaller than 0. So the validation does not do anything in this way. Let me check a little more.

**Comment by NickrenREN** [comment]
@jingxu97  Thanks for your comment.  Validation will return err when getNodeAllocatableAbsolute  sets the value to 0. I agree that we can call validateNodeAllocatable before  cm.setupNode(activePods) and call GetNodeAllocatableReservation with modification .
Codes changed. PTAL

**Comment by jingxu97** [comment]
Could you also add unit tests for this because we miss this part somehow before. Thank you!

**Comment by NickrenREN** [comment]
@jingxu97 TestNodeAllocatableInputValidation has tested this part I think in node_container_manager_test.go .  Correct me if I am wrong.

File: pkg/kubelet/cm/node_container_manager.go [hunk 1]
```diff
@@ -225,14 +226,26 @@ func hardEvictionReservation(thresholds []evictionapi.Threshold, capacity v1.Res
...
+		value, ok := capacityClone.(resource.Quantity)
+		if !ok {
+			return fmt.Errorf(
+				"failed to cast object %#v to Quantity",
+				capacityClone)
+		}
+		value.Sub(v)
+
+		if value.Sign() < 0 {
+			errors = append(errors, fmt.Sprintf("Resource %q has an allocatable of %v", k, v))
```
Comment (jingxu97) [inline pkg/kubelet/cm/node_container_manager.go]: maybe also put the capacity value in the error message.

**Comment by jingxu97** [comment]
@NickrenREN that test did not check if validation should fail as expected. That is why the test does not expose this bug earlier.

**Comment by jingxu97** [comment]
@NickrenREN could you also add a few more tests. (you don't need to change the existing ones)
1. Add ResourceStorageScratch in resource List. (I think you need to create a new func similar to getResourceList to include storage)
2. Include the test case that has EvictionThreashold

File: pkg/kubelet/cm/node_container_manager_test.go [hunk 2]
```diff
@@ -297,9 +297,74 @@ func TestNodeAllocatableInputValidation(t *testing.T) {
 			NodeConfig: nc,
 			capacity:   tc.capacity,
 		}
-		if err := cm.validateNodeAllocatable(); err != nil && !tc.invalidConfiguration {
```
Comment (NickrenREN) [inline pkg/kubelet/cm/node_container_manager_test.go]: I think the result check logic should be modified too. If `validateNodeAllocatable` returns no error, but we set the configuration to `invalid`, it should also log error

File: pkg/kubelet/cm/node_container_manager_test.go [hunk 3]
```diff
@@ -297,9 +297,74 @@ func TestNodeAllocatableInputValidation(t *testing.T) {
...
+		if err == nil && tc.invalidConfiguration {
+			t.Logf("Expected invalid node allocatable configuration")
+			t.FailNow()
+		} else if err != nil && !tc.invalidConfiguration {
 			t.Logf("Expected valid node allocatable configuration: %v", err)
 			t.FailNow()
 		}
 	}
+
+	storageEvictionThreshold := resource.MustParse("100Mi")
```
Comment (NickrenREN) [inline pkg/kubelet/cm/node_container_manager_test.go]: Sorry I did not notice the test cases before just checked `SignalMemoryAvailable`,  added `SignalNodeFsAvailable`

**Comment by NickrenREN** [comment]
@jingxu97  Done. Hope i have addressed your comments.

**Comment by jingxu97** [comment]
/lgtm
Thanks!

**Comment by derekwaynecarr** [comment]
i am unable to review deeply now, but i would expect local storage to work the same as memory.

can you confirm this keeps the behavior consistent as described for memory here:

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node-allocatable.md#allocatable

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node-allocatable.md#kubelet-evictions-thresholds

both resources support eviction, and i would expect the same invariants to hold.

**Comment by NickrenREN** [comment]
Yes, there is no difference.  Node allocatable = node capacity - reservation

GetNodeAllocatableReservation will get the sum of kubereservation, system reservation and eviction reservation.

**Comment by jingxu97** [comment]
@derekwaynecarr actually this fix is not specific to local storage. It is a bug in validateNodeAllocatable which validate all types of allocatable resources.
 
Could you please help approve?

**Comment by derekwaynecarr** [comment]
sorry for my confusion.

/approve no-issue
/lgtm

**Pull request merged**